### PR TITLE
[llvm][ARM] Emit MVE .arch_extension after .fpu directive if it does not include MVE features

### PR DESCRIPTION
**Pull request by simpal01**
The floating-point and MVE features together specify the MVE functionality that is supported on the Cortex-M85 processor. But the FPU extension for the underlying architecture(armv8.1-m.main) is FPV5 which does not include MVE-F. So Compiler's -S output and `-save-temps=obj` loses MVE feature which leads to assembler error. What happening here is .fpu directive overrides any previously set features by .cpu directive. Since the the corresponding .fpu generated (.fpu fpv5-d16) does not include MVE-F, it overrides those features even though it is supported and set by the .cpu directive. Looks like .fpu is supposed to do this.

In this case, there should be an .arch_extension directive re-enabling the relevant extensions after .fpu  if the goal is to keep these extensions enabled. GCC also does the same.

So this patch enables the MVE features by emitting the below arch extension:
  .fpu fpv5-d16
  .arch_extension mve.fp

### CODE REVIEW

**Comment by DavidSpickett** [comment]
I think the commit title would make more sense at a glance if it was saying what is being done instead of what's being fixed. How about:
```
[llvm][ARM] Emit MVE .arch_extension after .fpu directive if it does not include MVE features
```
Then the commit message is fine as is, and tells the reader why this was done and what it fixes.

File: clang/test/CodeGen/arm-v8.1m-check-mve.ll [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  %15 = load ptr, ptr %handle.addr, align 4
+  %d = getelementptr inbounds %struct.dummy_t, ptr %15, i32 0, i32 3
+  %16 = load float, ptr %d, align 4
+  %sub6 = fsub float 0x3F5439DE40000000, %16
+  %17 = load ptr, ptr %handle.addr, align 4
+  %d7 = getelementptr inbounds %struct.dummy_t, ptr %17, i32 0, i32 3
+  %18 = load float, ptr %d7, align 4
+  %19 = call float @llvm.fmuladd.f32(float 0x3FA47AE140000000, float %sub6, float %18)
+  store float %19, ptr %d7, align 4
+  ret i8 0
```
Comment (DavidSpickett) [inline clang/test/CodeGen/arm-v8.1m-check-mve.ll:53]: Do you need such a large example just to get the directives emitted?

File: clang/test/CodeGen/arm-v8.1m-check-mve.ll [hunk 2]
```diff
@@ -0,0 +1,56 @@
+; REQUIRES: arm-registered-target
+; RUN: %clang --target=arm-none-eabi -mcpu=cortex-m85 -mfloat-abi=hard -save-temps=obj -S -o - %s | FileCheck %s
+; RUN: %clang --target=arm-none-eabi -mcpu=cortex-m55 -mfloat-abi=hard -save-temps=obj -S -o - %s | FileCheck %s
+; RUN: %clang --target=arm-none-eabi -mcpu=cortex-m85 -mfloat-abi=hard -O2 -c -mthumb -save-temps=obj %s
+; RUN: %clang --target=arm-none-eabi -mcpu=cortex-m55 -mfloat-abi=hard -O2 -c -mthumb -save-temps=obj %s
+; CHECK: .fpu   fpv5-d16
+; CHECK  .arch_extension mve.fp
```
Comment (DavidSpickett) [inline clang/test/CodeGen/arm-v8.1m-check-mve.ll:7]: Is arch_extension expected to always be on the very next line? If so, CHECK-NEXT would be bit more robust.

File: clang/test/CodeGen/arm-v8.1m-check-mve.ll [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
+target triple = "thumbv8.1m.main-none-unknown-eabihf"
```
Comment (DavidSpickett) [inline clang/test/CodeGen/arm-v8.1m-check-mve.ll:9]: Not 100% sure but I think you can leave these out and they'll be got from the options given to clang regardless (not very important though).

**Comment by simpal01** [comment]
> I think the commit title would make more sense at a glance if it was saying what is being done instead of what's being fixed. How about:
> 
> ```
> [llvm][ARM] Emit MVE .arch_extension after .fpu directive if it does not include MVE features
> ```
> 
> Then the commit message is fine as is, and tells the reader why this was done and what it fixes.

Done

**Comment by simpal01** [comment]
> I think the commit title would make more sense at a glance if it was saying what is being done instead of what's being fixed. How about:
> 
> ```
> [llvm][ARM] Emit MVE .arch_extension after .fpu directive if it does not include MVE features
> ```
> 
> Then the commit message is fine as is, and tells the reader why this was done and what it fixes.

Done

File: clang/test/CodeGen/arm-v8.1m-check-mve.ll (hunk 3 above)
Comment (simpal01) [reply clang/test/CodeGen/arm-v8.1m-check-mve.ll]: Done

File: clang/test/CodeGen/arm-v8.1m-check-mve.ll (hunk 2 above)
Comment (simpal01) [reply clang/test/CodeGen/arm-v8.1m-check-mve.ll]: Done

File: clang/test/CodeGen/arm-v8.1m-check-mve.ll (hunk 1 above)
Comment (DavidSpickett) [reply clang/test/CodeGen/arm-v8.1m-check-mve.ll:51]: Ok I see, you need to setup the call to this fmuladd, that you know will emit MVE instructions.

In that case, could you add a check line for one of the MVE instructions as well? That would make the intent clearer, and warn us in the (unlikely) case that this code stopped emitting MVE instructions.

File: clang/test/CodeGen/arm-v8.1m-check-mve.c [hunk 4]
```diff
@@ -0,0 +1,35 @@
+// RUN: %clang --target=arm-none-eabi -mcpu=cortex-m85 -mfloat-abi=hard -O2 -save-temps=obj -S -o - %s | FileCheck %s
```
Comment (DavidSpickett) [inline clang/test/CodeGen/arm-v8.1m-check-mve.c:1]: Should there be an m55 test for this as well, as you've got 55 and 85 for save-temps.

**Comment by DavidSpickett** [comment]
Just small things on the test case from me. Nominate @davemgreen to give it a final check and approval, since I have not been around this area in a while.

File: clang/test/CodeGen/arm-v8.1m-check-mve.c [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+// REQUIRES: arm-registered-target
+
+// CHECK: .fpu   fpv5-d16
+// CHECK-NEXT  .arch_extension mve.fp
+
+#define DUMMY_CONST_1 (0.0012345F)
+#define DUMMY_CONST_2 (0.01F)
+#define DUMMY_CONST_3 (0.02F)
+#define DUMMY_CONST_4 (0.03F)
+#define DUMMY_CONST_5 (0.04F)
```
[the remark this answers is not in the document]
Comment (simpal01) [reply clang/test/CodeGen/arm-v8.1m-check-mve.c]: Done

File: clang/test/CodeGen/arm-v8.1m-check-mve.c [hunk 6]
```diff
@@ -0,0 +1,35 @@
+// RUN: %clang --target=arm-none-eabi -mcpu=cortex-m85 -mfloat-abi=hard -O2 -save-temps=obj -S -o - %s | FileCheck %s
+// RUN: %clang --target=arm-none-eabi -mcpu=cortex-m85 -mfloat-abi=hard -O2 -c -mthumb -save-temps=obj %s
+// RUN: %clang --target=arm-none-eabi -mcpu=cortex-m55 -mfloat-abi=hard -O2 -c -mthumb -save-temps=obj %s
```
[the remark this answers is not in the document]
Comment (simpal01) [reply clang/test/CodeGen/arm-v8.1m-check-mve.c:7]: Done

File: clang/test/CodeGen/arm-v8.1m-check-mve.c (hunk 4 above)
Comment (simpal01) [reply clang/test/CodeGen/arm-v8.1m-check-mve.c:1]: Done

**Comment by DavidSpickett** [comment]
Tests now LGTM, thanks for the updates.

File: clang/test/CodeGen/arm-v8.1m-check-mve.c [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+{
+    float a;
+    float b;
+    float c;
+    float d;
+} dummy_t;
+
+// CHECK-LABEL: foo
+// CHECK: vsub.f32
+// CHECK: vfma.f32
```
Comment (davemgreen) [inline clang/test/CodeGen/arm-v8.1m-check-mve.c:26]: This looks like it is trying to SLP vectorize to vector operations? These instructions will match both `vsub.f32 s6, s2, s10` and `vsub.f32 q6, q2, q1` operations though.

**Review by davemgreen — COMMENTED** [review]
Hi. From what I can tell the logic looks OK. We add the archextension in places we expect to now?

It is generally considered best practice to not have clang test that run the entire pass pipeline. In this case it looks like it's trying to SLP vectorize the code to make sure MVE operations are produced?  The problem can be that decision like that can change, and it is better if the tests are more narrowly focussed than replying on the whole pipeline. You do loose end-to-end testing but it may be better to have a clang test that checks the IR generated is what is expected, and an assembly test to make sure the `.arch_extension mve.fp` is recognized and turns on the MVE instructions it should do.

**Comment by simpal01** [comment]
> Hi. From what I can tell the logic looks OK. We add the archextension in places we expect to now?
> 
> It is generally considered best practice to not have clang test that run the entire pass pipeline. In this case it looks like it's trying to SLP vectorize the code to make sure MVE operations are produced? The problem can be that decision like that can change, and it is better if the tests are more narrowly focussed than replying on the whole pipeline. You do loose end-to-end testing but it may be better to have a clang test that checks the IR generated is what is expected, and an assembly test to make sure the `.arch_extension mve.fp` is recognized and turns on the MVE instructions it should do.

DONE

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM

**Pull request merged**